src/main/java/net/openhft/chronicle/queue/impl/single/StoreAppender.java

```diff
@@ -657,14 +717,14 @@ protected void writeBytesInternal(final long index, @NotNull final BytesStore by
             rollCycleTo(cycle, this.cycle > cycle);
 
         // in case our cached headerNumber is incorrect.
-        resetPosition(true);
+        resetPosition(indexValidationMode == IndexValidationMode.SAFE);
 
         long headerNumber = wire.headerNumber();
 
         boolean isNextIndex = index == headerNumber + 1;
         if (!isNextIndex) {
             if (index > headerNumber + 1)
-                throw new IllegalStateException("Unable to move to index " + Long.toHexString(index) + " beyond the end of the queue, current: " + Long.toHexString(headerNumber));
+                throw new IndexOutOfBoundsException("Unable to move to index " + Long.toHexString(index) + " beyond the end of the queue, current: " + Long.toHexString(headerNumber));
```

Comment:
This is potentially controversial.

While I doubt anyone was catching this `IllegalStateException`, throwing `IndexOutOfBoundsException` is an interface change but arguably a better fit for what's occurred here, and it also allows you to distinguish programmatically between the index failing validation (and no write being performed) and discovering we wrote at the wrong index AFTER the excerpt was written (as is checked below).

Reply:
I vote for a new subclass of IllegalStateException ([we use that pattern commonly](https://github.com/search?q=org%3AOpenHFT%20%22extends%20IllegalStateException%22&type=code))

Reply:
It really should be an IllegalArgumentException anyway, not an IllegalStateException

Reply:
I hear you. Could we change it to IAE in x.26?

Reply:
Well it is marked as an internal API, you really need to reach under the hood, or use `InternalAppender` interface to get at it.

Reply:
Yes fair point - it is an internal method. Lets change it, but I think IAE is safer than IndexOutOfBoundsE as our convention tends to throw IAE, ISE, or subclasses thereof

Reply:
Yeah cool, will do the specialise IAE